activesupport/test/event_reporter_test.rb

```diff
@@ -0,0 +1,696 @@
+      assert_equal([{ subscriber: @subscriber, filter: nil }], reporter.subscribers)
+    end
+
+    test "#subscribe with filter" do
+      reporter = ActiveSupport::EventReporter.new
+
+      filter = ->(event) { event[:name].start_with?("user.") }
+      reporter.subscribe(@subscriber, &filter)
+
+      assert_equal([{ subscriber: @subscriber, filter: filter }], reporter.subscribers)
+    end
+
+    test "#subscribe raises ArgumentError when sink doesn't respond to emit" do
+      invalid_subscriber = Object.new
+
+      error = assert_raises(ArgumentError) do
+        @reporter.subscribe(invalid_subscriber)
+      end
+
+      assert_equal "Event subscriber Object must respond to #emit", error.message
+    end
+
+    test "#notify with name" do
+      assert_called_with(@subscriber, :emit, [
+        event_matcher(name: "test_event")
+      ]) do
+        @reporter.notify(:test_event)
+      end
+    end
+
+    test "#notify filters" do
+      reporter = ActiveSupport::EventReporter.new
+      reporter.subscribe(@subscriber) { |event| event[:name].start_with?("user_") }
+
+      assert_not_called(@subscriber, :emit) do
+        reporter.notify(:test_event)
+      end
+
+      assert_called_with(@subscriber, :emit, [
+        event_matcher(name: "user_event")
+      ]) do
+        reporter.notify(:user_event)
+      end
+    end
+
+    test "#notify with name and hash payload" do
+      assert_called_with(@subscriber, :emit, [
+        event_matcher(name: "test_event", payload: { key: "value" })
+      ]) do
+        @reporter.notify(:test_event, { key: "value" })
+      end
+    end
+
+    test "#notify with name and kwargs" do
+      assert_called_with(@subscriber, :emit, [
+        event_matcher(name: "test_event", payload: { key: "value" })
+      ]) do
+        @reporter.notify(:test_event, key: "value")
+      end
+    end
+
+    test "#notify symbolizes keys in hash payload" do
+      assert_called_with(@subscriber, :emit, [
+        event_matcher(name: "test_event", payload: { key: "value" })
+      ]) do
+        @reporter.notify(:test_event, { "key" => "value" })
+      end
+    end
+
+    test "#notify with hash payload and kwargs raises" do
+      error = assert_raises(ArgumentError) do
+        @reporter.notify(:test_event, { key: "value" }, extra: "arg")
+      end
+
+      assert_match(
+        /Rails.event.notify accepts either an event object, a payload hash, or keyword arguments/,
+        error.message
+      )
+    end
+
+    test "#notify includes source location in event payload" do
+      def test_method
+        @reporter.notify("test_event")
+      end
+      lineno = __LINE__ - 2
+
+      assert_called_with(@subscriber, :emit, [
+        event_matcher(name: "test_event", source_location: {
+          filepath: __FILE__,
+          lineno: lineno
+        })
+      ]) do
+        test_method
+      end
+    end
+
+    test "#notify with caller depth option" do
+      def custom_log_abstraction(message)
```

Comment:
Method defined inside a test is strange. It will leak to other tests. I'd just define a custom class outside of the test.